src/renderers/shared/reconciler/ReactCompositeComponent.js

```diff
@@ -168,7 +172,28 @@ var ReactCompositeComponentMixin = {
         inst = new Component(publicProps, publicContext, ReactUpdateQueue);
       }
     } else {
-      inst = new StatelessComponent(Component);
+      if (__DEV__) {
+        ReactCurrentOwner.current = this;
+        try {
+          inst = Component(publicProps, publicContext, ReactUpdateQueue);
+        } finally {
+          ReactCurrentOwner.current = null;
+        }
+      } else {
+        inst = Component(publicProps, publicContext, ReactUpdateQueue);
+      }
+      if (inst == null || inst.render == null) {
+        renderedElement = inst;
+        warnIfInvalidElement(Component, renderedElement);
+        invariant(
+          inst === null ||
+          inst === false ||
+          ReactElement.isValidElement(inst),
+          '%s(...): Stateless function components must return a valid ReactElement.',
```

Comment:
Can we have more consistent wording here and in _renderValidatedComponent? (Looks like that one's currently wrong because it says component instead of element.) Maybe:

> A valid React element (or null) must be returned. You may have returned undefined, an array or some other invalid object.